### PR TITLE
Pandaria and Draenor Scripts

### PR DESCRIPTION
**Pull request by Infamous-devel**
**Changes proposed**:
- Reorganization of Script Folders for Pandaria and Draenor Scripts.
- Added ALL Draenor Bosses. (Spell, Events, Summons).
- Added ALL Pandaria Bosses. (Spell, Events, Summons, Phases).
- Added Pandaria Zones. (Creatures, AreaTriggers, ItemScripts, ...)
- Added Bosses Summon NPCs.
- Added Bosses NPC CreatureScripts.
- Added SQL with creature_template UPDATES (ScriptName).

**Target branch(es)**: 6.x

**Tests performed**: (Does it build, tested in-game, etc)
- Compile Without Errors
- Bosses Tested In-Game (Some of them) [ONLY if Spells are casted] 

**Known issues and TODO list**:
- [x] Pandaria Bosses
- [x] Pandaria Zones
- [x] Draenor Bosses
- [ ] Draenor Zones
- [x] Instances.
- [x] Bosses 80% Blizzlike.
- [ ] Bosses Spawns not Added.
- [ ] Instances not Populated.
- [ ] Some Functions Commented... (Not sure how to rewrite).
- [x] Bosses doesn't have Phases Coded.
- [x] Bosses doesn't have Emotes and Says Coded.
- [x] Bosses summons creatures. (If blizzlike call nearby creatures).
- [x] Some Issues with MMaps (Some Boss disappears when specific Spell is casted??).
- [x] Compile errors due to latest commits ( undefined reference to in ScriptLoader).

### CODE REVIEW

**Comment by Kittnz** [comment]
Tabs, tabs everywhere

**Comment by Aokromes** [comment]
Why delete  the whole existing files to add one or two rows?
Also, quite sure the scripts don't needs the whole amount of includes you add.

**Comment by Infamous-devel** [comment]
Kittnz: I used Eclipse IDE AutoTab.

Aokromes: That's my fault. I did something I should not.

**Comment by Aokromes** [comment]
Also, mangos have draenor scripted?

**Comment by Infamous-devel** [comment]
> Also, mangos have draenor scripted?

I don't think so..

**Comment by Aokromes** [comment]
Then remove mangos (C) :)

**Comment by Kittnz** [comment]
https://github.com/TrinityCore/TrinityCore/pull/16793/files#diff-a23309cb3f904f2d11f9d91fd53ef349R3

File: src/server/scripts/Custom/custom_script_loader.cpp [hunk 1]
```diff
@@ -17,9 +17,9 @@
 
 // This is where scripts' loading functions should be declared:
```
Comment (Aokromes) [inline src/server/scripts/Custom/custom_script_loader.cpp]: remove this change from pr.

File: src/server/scripts/Custom/custom_script_loader.cpp [hunk 2]
```diff
@@ -17,9 +17,9 @@
...
 
-
 // The name of this function should match:
 // void Add${NameOfDirectory}Scripts()
 void AddCustomScripts()
 {
+
```
Comment (Aokromes) [inline src/server/scripts/Custom/custom_script_loader.cpp]: remove this change from pr.

File: sql/Draenor/some_draenor_bosses.sql [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+-- Auchindoun
+
+UPDATE creature_template SET minlevel=97, maxlevel=103, faction=1791, ScriptName='boss_vigilant_kaathar' WHERE entry=75839;
+UPDATE creature_template SET minlevel=97, maxlevel=103, faction=1791, ScriptName='boss_soulbinder_nyami' WHERE entry=76177;
+UPDATE creature_template SET minlevel=97, maxlevel=103, faction=1791, ScriptName='boss_azzakel' WHERE entry=75927;
+UPDATE creature_template SET minlevel=97, maxlevel=103, faction=1791, ScriptName='boss_terongor' WHERE entry=77734;
+
+-- Sky Reach
+
+UPDATE creature_template SET minlevel=99, maxlevel=103, faction=1791, ScriptName='boss_ranjit' WHERE entry=75964;
```
Comment (Aokromes) [inline sql/Draenor/some_draenor_bosses.sql]: No new line. Non-standard filename.

**Comment by r00ty-tc** [comment]
Regarding tabs, VS also does this by default. You can change it to spaces. TC convention is spaces and not tabs and it makes things very messy if we all use our own codestyle and tab style.

**Comment by Aokromes** [comment]
He looks to be using eclipse.

**Comment by Infamous-devel** [comment]
Yes I do. So I'm changing the tabs to spaces now,

**Comment by r00ty-tc** [comment]
Hence "also". Surely Eclipse also has a similar setting?

File: src/server/scripts/Draenor/SkyReach/boss_ranjit.cpp [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "ScriptMgr.h"
```
Comment (Aokromes) [inline src/server/scripts/Draenor/SkyReach/boss_ranjit.cpp]: are you sure this huge amount of includes are needed? The same for the rest.

**Comment by Infamous-devel** [comment]
Aokromes: Aren't needed at All but when I tried to include ScriptPCH.h that includes all that give me an error when I compile. So I decide to wrote whole includes there.

**Comment by Aokromes** [comment]
Add only the needed only, adding to much only increases compile time.

**Comment by Naios** [comment]
I personally vote for using the `TaskScheduler` instead of the `EventMap` for new scripts. The code looks much cleaner then and you don't need to repeat yourself on many places.

**Comment by Infamous-devel** [comment]
Aokromes: I'll check the includes now.
Naios: Would try to code future scripts with TaskScheduler.

File: src/server/scripts/Draenor/Auchindoun/boss_soulbinder_nyami.cpp [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+{
+    SPELL_MIND_SPIKE = 154415,
+    SPELL_SHADOW_WORD_PAIN = 154477,
+    SPELL_SOUL_VESSEL = 153477,
+    SPELL_TORN_SPIRITS = 153994,
+};
+
+enum Events
+{
+    EVENT_MIND_SPIKE,
```
Comment (jackpoz) [inline src/server/scripts/Draenor/Auchindoun/boss_soulbinder_nyami.cpp]: https://ideone.com/cUzTMl
You claim you tested all bosses in-game, yet you didn't notice that EVENT_MIND_SPIKE is never triggered.
Testing a script means adding a breakpoint to every event and ensuring it's executed.

**Comment by Infamous-devel** [comment]
jackpoz: As I wrote in the description, the bosses aren't 100% Blizzlike. Their behavior isn't the same, I mean... the bosses are coded in less than 60% of their originals...

**Comment by jackpoz** [comment]
it's broken C++ code, it has nothing to do with being blizzlike or not.

**Comment by Infamous-devel** [comment]
jackpoz: You're right. I forgot to put the value of the first enum...  :(
I'm fixing now.

**Comment by ShinDarth** [comment]
@Infamous-devel  nice initiative!

**Comment by Exodius** [comment]
Great effort overall, but... Wouldn't it be easier, for yourself and those that will review the code, to do this Dungeon by Dungeon, Boss by Boss?

**Comment by Infamous-devel** [comment]
@Exodius It would but, I wrote that code for my server time ago, and now I'm pulling all that I have..

**Comment by MitchesD** [comment]
please fix codestyle in these files, tabs etc. its not readable

**Comment by Infamous-devel** [comment]
@MitchesD I replace Tabs to Spaces in last commits and changed the codestyle too. Where I did'nt fix it?

**Comment by MitchesD** [comment]
@Infamous-devel everywhere? http://i.imgur.com/2lNGiUi.png

**Comment by Infamous-devel** [comment]
@MitchesD I guess that I fixed it.

File: sql/Draenor/2016_03_16_00_world.sql [hunk 6]
```diff
@@ -0,0 +1,23 @@
+--- Bloodmaul Slag Mines
```
Comment (dantewow) [inline sql/Draenor/2016_03_16_00_world.sql]: You have to make a correction on Line 1.
You have a syntax error.
Line 1: - - - ( you should have only 2 - - not 3 - - - )
You are not able to import the sql on database cause of line 1. 
This is the correct syntax: -- Bloodmaul Slag Mines

**Comment by DDuarte** [comment]
@Infamous-devel: I discussed this PR with someone else on IRC and the consensus is that we wished that you focused on a single boss at a time, instead of submitting a mega PR with bosses working 10%.  
A single PR that scripts a single boss, actually researched, sniffed and tested is immensely more valuable than this current PR.

**Comment by DevQuad** [comment]
I'm more surprised at the whole admin composition TC. People do pull that would in any way help the develop Draenor in terms of gaming component even those 10%. You manage to deflect and help here. Over time, it will be possible to modify all.

**Comment by DDuarte** [comment]
I'm sorry if I sounded more negative than intended, I didn't mean to discourage you. My whole point is that you should focus on doing one thing at a time, e.g, by submitting one boss or one dungeon per PR.

P.S I'm not sure if I understood everything that you have written.

File: src/server/scripts/Draenor/Auchindoun/boss_soulbinder_nyami.cpp [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    {
+        return new boss_soulbinder_nyamiAI(creature);
+    }
+};
+
+void AddSC_boss_soulbinder_nyami()
+{
+    new boss_soulbinder_nyami();
+}
+
```
Comment (tkrokli) [inline src/server/scripts/Draenor/Auchindoun/boss_soulbinder_nyami.cpp]: Blank line, redundant.

File: src/server/scripts/Draenor/Auchindoun/boss_terongor.cpp [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+    {
+        return new boss_terongorAI(creature);
+    }
+};
+
+void AddSC_boss_terongor()
+{
+    new boss_terongor();
+}
+
```
Comment (tkrokli) [inline src/server/scripts/Draenor/Auchindoun/boss_terongor.cpp]: Blank line, redundant.

File: src/server/scripts/Draenor/BloodmaulSlagMines/boss_gugrokk.cpp [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+};
+
+void AddSC_Boss_Gugrokk()
+{
+    new Boss_Gugrokk();
+    new NPC_Unstable_Slag();
+}
+
```
Comment (tkrokli) [inline src/server/scripts/Draenor/BloodmaulSlagMines/boss_gugrokk.cpp]: 3 blank lines, redundant.

File: src/server/scripts/Draenor/BloodmaulSlagMines/boss_magmolatus.cpp [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    {
+        return new Boss_MagmolatusAI(creature);
+    }
+};
+
+void AddSC_Boss_Magmolatus()
+{
+    new Boss_Magmolatus();
+}
+
```
Comment (tkrokli) [inline src/server/scripts/Draenor/BloodmaulSlagMines/boss_magmolatus.cpp]: 2 blank lines, redundant.

File: src/server/scripts/Draenor/IronDocks/boss_ahriok_dugru.cpp [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    {
+        return new boss_ahriok_dugruAI(creature);
+    }
+};
+
+void AddSC_boss_ahriok_dugru()
+{
+    new boss_ahriok_dugru();
+}
+
```
Comment (tkrokli) [inline src/server/scripts/Draenor/IronDocks/boss_ahriok_dugru.cpp]: 8 blank lines, redundant.

File: src/server/scripts/Draenor/SkyReach/boss_ranjit.cpp [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+    {
+        return new boss_ranjitAI(creature);
+    }
+};
+
+void AddSC_boss_ranjit()
+{
+    new boss_ranjit();
+}
+
```
Comment (tkrokli) [inline src/server/scripts/Draenor/SkyReach/boss_ranjit.cpp]: 4 blank lines, redundant.

File: src/server/scripts/Draenor/draenor_script_loader.cpp [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+	AddSC_boss_vigilant_kaathar();
+	AddSC_boss_soulbinder_nyami();
+	AddSC_boss_azzakel();
+	AddSC_boss_terongor();
+
+	/* Sky Reach */
+	AddSC_boss_ranjit();
+
+}
+
```
Comment (tkrokli) [inline src/server/scripts/Draenor/draenor_script_loader.cpp]: 4 blank lines, redundant.

File: src/server/scripts/Draenor/draenor_script_loader.cpp [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+void AddSC_boss_vigilant_kaathar();
+void AddSC_boss_soulbinder_nyami();
+void AddSC_boss_azzakel();
+void AddSC_boss_terongor();
+
+void AddSC_boss_ranjit();
+
+void AddDraenorScripts()
+{
+
```
Comment (tkrokli) [inline src/server/scripts/Draenor/draenor_script_loader.cpp]: Blank line, redundant.

File: src/server/scripts/Draenor/draenor_script_loader.cpp [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+	/* Auchindoun */
+	AddSC_boss_vigilant_kaathar();
+	AddSC_boss_soulbinder_nyami();
+	AddSC_boss_azzakel();
+	AddSC_boss_terongor();
+
+	/* Sky Reach */
+	AddSC_boss_ranjit();
+
```
Comment (tkrokli) [inline src/server/scripts/Draenor/draenor_script_loader.cpp]: Blank line, redundant.

File: src/server/scripts/Pandaria/pandaria_script_loader.cpp [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+void AddPandariaScripts()
+{
+}
+
```
Comment (tkrokli) [inline src/server/scripts/Pandaria/pandaria_script_loader.cpp]: 2 blank lines, redundant.

**Comment by KarrokDC** [comment]
Awesome to have someone start work on MoP/WoD! :+1:  After laying down the foundation, perhaps do PR's per boss as suggested, maybe others can hook into your starting structure and fully flesh out a boss too :)

**Comment by Infamous-devel** [comment]
The idea is to order the new Pandaria and Draenor Scripts into subfolder as all the other scripts. Besides I'm writing the Boss templates to make easier developers write a full Blizz code.

**Comment by Infamous-devel** [comment]
Can't compile because of errors of latest commits...

**Comment by cooler-SAI** [comment]
Good JOB Here! It's good news that someone work on WOD and Panda Content! Keep up! and THANKS again!

File: .gitignore [hunk 17]
```diff
@@ -17,3 +17,5 @@ CMakeLists.txt.user
 *.BASE.*
 *.LOCAL.*
 nbproject/*
+.project
+.cproject
```
Comment (Aokromes) [inline .gitignore]: revert this change.

File: src/server/game/AuctionHouseBot/AuctionHouseBotSeller.cpp [hunk 18]
```diff
@@ -547,7 +547,7 @@ void AuctionBotSeller::LoadItemsQuantity(SellerConfiguration& config)
     config.SetRandomStackRatioPerClass(ITEM_CLASS_QUIVER, sAuctionBotConfig->GetConfig(CONFIG_AHBOT_CLASS_RANDOMSTACKRATIO_QUIVER));
     config.SetRandomStackRatioPerClass(ITEM_CLASS_QUEST, sAuctionBotConfig->GetConfig(CONFIG_AHBOT_CLASS_RANDOMSTACKRATIO_QUEST));
     config.SetRandomStackRatioPerClass(ITEM_CLASS_KEY, sAuctionBotConfig->GetConfig(CONFIG_AHBOT_CLASS_RANDOMSTACKRATIO_KEY));
-    config.SetRandomStackRatioPerClass(ITEM_CLASS_MISC, sAuctionBotConfig->GetConfig(CONFIG_AHBOT_CLASS_RANDOMSTACKRATIO_MISC));
+    config.SetRandomStackRatioPerClass(ITEM_CLASS_MISCELLANEOUS, sAuctionBotConfig->GetConfig(CONFIG_AHBOT_CLASS_RANDOMSTACKRATIO_MISC));
     config.SetRandomStackRatioPerClass(ITEM_CLASS_GLYPH, sAuctionBotConfig->GetConfig(CONFIG_AHBOT_CLASS_RANDOMSTACKRATIO_GLYPH));
```
Comment (Aokromes) [inline src/server/game/AuctionHouseBot/AuctionHouseBotSeller.cpp]: The fix of compile don't belongs to this pr.

File: src/server/scripts/CMakeLists.txt [hunk 19]
```diff
@@ -1,79 +1,79 @@
-# Copyright (C) 2008-2016 TrinityCore <http://www.trinitycore.org/>
```
Comment (Aokromes) [inline src/server/scripts/CMakeLists.txt]: Why you change this?

**Comment by Infamous-devel** [comment]
@Aokromes I know that change don't belongs to this pr but can't compile the core without that fix.

**Comment by soulfrost** [comment]
all things which u call " scripts " seems like a shit.
write 40 lines of code and let bosses cast 4-5 spells.... omg - it's below 5% of all work on boss

**Comment by Infamous-devel** [comment]
At leats I have done more than you have... If you want to complain it do it, but not here. As I say in the desc and comments there are template of bosses. Don't comment shits. @soulfrost

**Comment by Infamous-devel** [comment]
"undefined reference to" Scripts... not sure how to fix it :(

**Comment by soulfrost** [comment]
@Infamous-devel  u dont have make more that i have... u'll never do this. at least u dont understand how to make this

**Comment by Kittnz** [comment]
@soulfrost No body care's what you have, it's not in a pr so we don't care.

**Comment by soulfrost** [comment]
@Kittnz it's just answer of his ``` At leats I have done more than you have ```
work which he want TODO should be dont part by part for one dungeon, not like shit for all dungeons - timers not blizzlike, no spell scripts ( which are needed ), no sql data, no right instance scripts etc etc etc

**Comment by Naios** [comment]
@soulfrost Seriously at least he publishes the work he has.... which has a higher benefit then some `shiny scripts in a private repo`.
If you really want to help here go on and publish the work you have.

**Comment by Kittnz** [comment]
We share, why  shouldn't you?
But you do have a point, and this is been said before by DDuarte. It's better to make PR's for each boss.

**Comment by soulfrost** [comment]
so... if DDurate allready said this - why this shitty PR still not closed and not made in right way?

p.s. not all of the code is mine personally, andi do not have the right to distribute it (nondisclosure agreement).

**Comment by Infamous-devel** [comment]
captura


Still getting that erro. Probably makefiles wrong?

File: src/server/scripts/CMakeLists.txt [hunk 20]
```diff
@@ -25,16 +25,18 @@ PrepareScripts(Spells PRIVATE_SOURCES)
...
-  PrepareScripts(World PRIVATE_SOURCES)
-  PrepareScripts(OutdoorPvP PRIVATE_SOURCES)
-  PrepareScripts(EasternKingdoms PRIVATE_SOURCES)
-  PrepareScripts(Kalimdor PRIVATE_SOURCES)
-  PrepareScripts(Outland PRIVATE_SOURCES)
-  PrepareScripts(Northrend PRIVATE_SOURCES)
-  PrepareScripts(Maelstrom PRIVATE_SOURCES)
-  PrepareScripts(Events PRIVATE_SOURCES)
-  PrepareScripts(Pet PRIVATE_SOURCES)
+  PrepareScripts(Custom scripts_STAT_SRCS)
```
Comment (Naios) [inline src/server/scripts/CMakeLists.txt]: You need to drop this changes to fix your build

**Comment by Infamous-devel** [comment]
Mandame un mail a: infamous.devel@gmail.com y hablamos.

**Comment by LuigiElleBalotta** [comment]
Do you need sniffs for some Draenor or Pandaria instances?

**Comment by Infamous-devel** [comment]
Would be fine if can send me some sniffs to complete code. I actually have most (90%) of Pandaria Raids, Dungeons and Zones Scripted (No 100% by me). I take the code from other emus based in older versions of TrinityCore and rewrited some parts making code compatible with updated versions. Some parts of the code are commented because I don't really know how to fix it.

**Comment by LuigiElleBalotta** [comment]
Contact me to my email on my profile

**Comment by DDuarte** [comment]
I would like to know how do you expect someone to review a pull request with 10k new lines of code and 90 commits named "Fix".

**Comment by Infamous-devel** [comment]
@DDuarte I know that is very hard to know what I'm doing in each commit with that comments. But that code was not expected to be in that pull request. Besides in that commits I was fixing code while I was compiling an receiving compile errors... Would comments better the commits in future updates.

**Comment by Keader** [comment]
@Infamous-devel why you dont close this PR and open one for each boss / script?
Would be easier to give you feedback and make revision easy.

**Comment by Infamous-devel** [comment]
@Keader Okay I'm closing the PR now.

**Comment by soulfrost** [comment]
roooooooooofl, ctrlc + ctrlv from public repos not blizzlike?

**Comment by DDuarte** [comment]
First (smaller) PR opened at #17011
Thanks Infamous for the decision :)
